### ISSUE TITLE
yield, bad_function call segfault if no callback and status

### ISSUE THREAD

**Issue by AmarOk1412** [Corvusoft/restbed#269]
Hi.

I think I found an issue. I use restbed to stream JSON objects. I have something like:

```cpp
s->yield(restbed::OK, "", [=]( const std::shared_ptr< restbed::Session > s) {
  foo->get(bar, [s](JSON) {
    //  Retreive JSON objects
    s->yield(JSON);
  }, [s]() {
    // Communication is finished
    s->close();
  });
});
```

But this version crash on yield because of a `bad_function call`. I found this issue (https://github.com/Corvusoft/restbed/issues/62), and this code works:

```cpp
s->yield(restbed::OK, "", [=]( const std::shared_ptr< restbed::Session > s) {
  foo->get(bar, [s](JSON) {
    //  Retreive JSON objects
    s->yield(JSON, [](const std::shared_ptr<restbed::Session> /*session*/){ }); // see the useless callback here. It solves everything :D
  }, [s]() {
     // Communication is finished
     s->close();
  });
});
```

For simplicity, this will segfault:
```cpp
s->yield(restbed::OK, "", [=]( const std::shared_ptr< restbed::Session > s) {
     for (auto i = 0; i < 5; ++i)
           s->yield("dummy");
     s->close();
});
```

Note: This code works (I think, didn't try):
```cpp
s->yield(restbed::OK, "", [=]( const std::shared_ptr< restbed::Session > s) {
     for (auto i = 0; i < 5; ++i)
           s->yield(restbed::OK, "dummy");
     s->close();
});
```

I test the code on some linux distrib.

Have a nice day,

**Comment by ben-crowhurst** [comment]
Thanks @AmarOk1412 we are really busy at the moment and will investigate ASAP.

**Comment by ben-crowhurst** [comment]
@AmarOk1412  I've attempted to replicate with the following test case.  No error was encountered.  Could you please provide a stacktrace?

``` C++
void get_handler( const shared_ptr< Session > session )
{
    session->yield( OK, "", [ ]( const shared_ptr< Session > session )
    {
        for ( auto i = 0; i < 5; ++i) session->yield( "dummy" );
        session->close( );
    } );
}

TEST_CASE( "yield callback leads to service crash", "[service]" )
{
    auto resource = make_shared< Resource >();
    resource->set_path( "test" );
    resource->set_method_handler( "GET", get_handler );
    
    auto settings = make_shared< Settings >();
    settings->set_port( 1984 );
    
    shared_ptr< thread > worker = nullptr;
    
    Service service;
    service.publish( resource );
    service.set_ready_handler( [&worker]( Service & service )
    {
        worker = make_shared< thread >( [&service]()
        {
            auto request = make_shared< Request >();
            request->set_method( "GET" );
            request->set_port( 1984 );
            request->set_host( "localhost" );
            request->set_path( "/test" );
            
            auto response = Http::sync( request );
            
            REQUIRE( 200 == response->get_status_code() );
            
            service.stop( );
        } );
    } );
    service.start( settings );
    worker->join();
}
```

**Comment by AmarOk1412** [comment]
I'll do it ASAP

**Comment by AmarOk1412** [comment]
@ben-crowhurst ok. I just retried and... doesn't segfault anymore. So seems to be fixed :). I close this.

**Comment by vortarian** [comment]
I was experiencing a similar issue, ran across this while debugging.  I found 2 situations that were contributing:
1) Calling yield repeatedly outside of callback (I was writing headers sequentially before starting to stream data recursively.)
2) Using recursive lambda's - I believe my stack objects were going out of scope even though I had specified the objects be passed through to the callback, the segfault occurred after exit of the function scope, but during the recursion.
3) My libraries were built with gcc-5.4, but the program was built with gcc-73..

After I resolved these 3 things I stopped having issues with stability.
